### PR TITLE
[WIP] e2e: async communication between test and callbacks

### PR DESCRIPTION
**Pull request by humblec**
The intention here is to add a framework  which looks similar to the  design in below doc:

https://hackmd.io/oIgJlaC3SaOG4xEaLV2K4w

Additional thoughts or validation of above is invited before taking this PR to next step.  

Cc @nixpanic @Madhu-1 @Yuggupta27 @yati1998  @Rakshith-R 

Just to give some example change in the caller ( once the implementation is done)  createsnapshot for 10 copies after changing the name and namespace becomes:

```
rc := RunConfig{{ {Name: createSnapshot, Args: *snaptemplate, 2}, Copies: 5, WforAll: true, PrepFuncs: [
      {{ SetName, SetNamespace}, arg {*snaptemplate}},....},
        
```

Then rc.Run() has been called, which internally spawn 10 routines to take care of the job and wait for all to finish and return the error or success back to the test case.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

### CODE REVIEW

**Comment by humblec** [comment]
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/upgrade-tests-cephfs/detail/upgrade-tests-cephfs/2113/pipeline
```! StartHost failed, but will try again: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute

* Creating kvm2 VM (CPUs=8, Memory=14336MB, Disk=20000MB) ...

* Failed to start kvm2 VM. Running "minikube delete" may fix it: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute


X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute
```

**Comment by humblec** [comment]
@nixpanic @Madhu-1 can you please  take a quick look at the design shared in the hackmd, so that I can start on implementation ? please feel free to propose any changes in the same hackmd

**Comment by humblec** [comment]
Considering no feedback, can I go ahead with the implementation @nixpanic   @Madhu-1 ?

**Comment by Madhu-1** [comment]
Let's discuss in today's standup the priority of this one.

**Comment by humblec** [comment]
> Let's discuss in today's standup the priority of this one.

@Madhu-1 I was mainly asking any comment on the design rather than the priority side :)

**Comment by Madhu-1** [comment]
> > Let's discuss in today's standup the priority of this one.
> 
> @Madhu-1 I was mainly asking any comment on the design rather than the priority side :)

@humblec Sorry I didn't find time to review this one.

File: e2e/async.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+type objects struct {
+	snap *snapapi.VolumeSnapshot
+	pvc  *v1.PersistentVolumeClaim
+	app  *v1.Pod
+}
+
+func executeOperations(ch <-chan configItem) {
+	for proc := range ch {
+		switch proc.callbackFn {
```
Comment (nixpanic) [inline e2e/async.go:29]: `callbackFn` should probably just be a `func()` and get executed. Coding of the async operations should all be at the same place where the test is executed, not spread out over different files.

**Pull request closed without merging**